src/main/scala/workflow/Pipeline.scala

```diff
@@ -0,0 +1,197 @@
+    val newDataDeps = this.dataDeps.map(_.map {
+      x => if (x == Pipeline.SOURCE) this.nodes.size + 1 else x
+    }) :+ Seq(this.sink) :+ Seq() :+ Seq(Pipeline.SOURCE)
+    val newFitDeps = this.fitDeps :+ Seq() :+ Seq() :+ Seq(this.nodes.size)
+    val newSink = newNodes.size - 1
+
+    val fittedTransformer = Pipeline[B, C](newNodes, newDataDeps, newFitDeps, newSink)
+    val totalOut = this andThen fittedTransformer
+    new PipelineWithFittedTransformer(totalOut.nodes, totalOut.dataDeps, totalOut.fitDeps, totalOut.sink, fittedTransformer)
+  }
+
+  final def andThen[C, L](est: LabelEstimator[B, C, L], data: RDD[A], labels: RDD[L]): PipelineWithFittedTransformer[A, B, C] = {
+    val transformerLabel = est.label + ".fit"
+
+    val newNodes = this.nodes :+ est :+ DataNode(data) :+ DataNode(labels) :+ new DelegatingTransformer[C](transformerLabel)
+    val newDataDeps = this.dataDeps.map(_.map {
+      x => if (x == Pipeline.SOURCE) this.nodes.size + 1 else x
+    }) :+ Seq(this.sink, this.nodes.size + 2) :+ Seq() :+ Seq() :+ Seq(Pipeline.SOURCE)
+    val newFitDeps = this.fitDeps :+ Seq() :+ Seq() :+ Seq() :+ Seq(this.nodes.size)
+    val newSink = newNodes.size - 1
+
+    val fittedTransformer = Pipeline[B, C](newNodes, newDataDeps, newFitDeps, newSink)
+    val totalOut = this andThen fittedTransformer
+    new PipelineWithFittedTransformer(totalOut.nodes, totalOut.dataDeps, totalOut.fitDeps, totalOut.sink, fittedTransformer)
+  }
+
+
+  final def toDOTString: String = {
+    val nodeLabels: Seq[String] = "-1 [label=\"In\" shape=\"Msquare\"]" +: nodes.zipWithIndex.map {
+      case (data: DataNode, id)  => s"$id [label=${'"' + data.label + '"'} shape=${"\"box\""} style=${"\"filled\""}]"
+      case (transformer: TransformerNode[_], id) => s"$id [label=${'"' + transformer.label + '"'}]"
+      case (estimator: EstimatorNode, id) => s"$id [label=${'"' + estimator.label + '"'} shape=${"\"box\""}]"
+    } :+ s"${nodes.size} [label=${"\"Out\""} shape=${"\"Msquare\""}]"
+
+    val dataEdges: Seq[String] = dataDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"$x -> $id")
+    } :+ s"$sink -> ${nodes.size}"
+
+    val fitEdges: Seq[String] = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"$x -> $id [dir=${"\"none\""} style=${"\"dashed\""}]")
+    }
+
+    val ranks = fitDeps.zipWithIndex.flatMap {
+      case (deps, id) => deps.map(x => s"{rank=same; $x $id}")
+    }
+
+    val lines = nodeLabels ++ dataEdges ++ fitEdges ++ ranks
+    lines.mkString("digraph pipeline {\n  rankdir=LR;\n  ", "\n  ", "\n}")
+  }
+
+  final private[workflow] def planEquals(pipeline: Pipeline[A, B]): Boolean = {
+    this.eq(pipeline) || (
+        (nodes == pipeline.nodes) &&
+        (dataDeps == pipeline.dataDeps) &&
+        (fitDeps == pipeline.fitDeps) &&
+        (sink == pipeline.sink))
+  }
+}
+
+object Pipeline {
+  val SOURCE: Int = -1
```

Comment:
This very well may have been me overfitting DSL stuff into the core DAG representation, but like you mention I can't think of any cases where the final pipeline doesn't have a single input.